### PR TITLE
Simplify AtomicFileWriter and use clearer temporary file names

### PR DESCRIPTION
**Pull request by Vlatombe**
* Uses a more readable name for the temporary file, derived from the target path
* Streamline the fallback logic for atomic move, and just let the error bubble up otherwise rather than logging then throwing which was leading to duplicate stacktraces in the system logs
* Record lack of atomic move support, falling back to replace existing (non-atomic) implementation for the current call as well as for future calls.

Related to https://github.com/jenkinsci/jenkins/pull/2548#discussion_r79500936

See [JENKINS-XXXXX](https://issues.jenkins.io/browse/JENKINS-XXXXX).

### Proposed changelog entries

- Avoid printing the same stacktrace multiple times when file persistence fails. Temporary files names used by the atomic file writer are now derived from the target file name.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/util/AtomicFileWriter.java
Comment (jglick) [inline core/src/main/java/hudson/util/AtomicFileWriter.java:157]: or simply
```suggestion
            tmpPath = File.createTempFile(destPath.getFileName() + "-atomic", "tmp", dir.toFile()).toPath();
```

File: core/src/main/java/hudson/util/AtomicFileWriter.java
Comment (jglick) [inline core/src/main/java/hudson/util/AtomicFileWriter.java:243]:
```suggestion
                Files.move(source, destination, StandardCopyOption.ATOMIC_MOVE);
                return;
```
to avoid a race condition if another thread just turned this flag off.

**Review by jglick — APPROVED** [review]

**Review by jglick — APPROVED** [review]

**Review by timja — APPROVED** [review]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by jglick** [comment]
@timja @Vlatombe I think this deserves a changelog entry? IIUC this has an observable benefit to Jenkins on certain filesystems, at least in terms of the quality of log output.

**Comment by Vlatombe** [comment]
I updated the description with a proposed changelog entry.

**Review by Thompson1985 — APPROVED** [review]

**Comment by basil** [comment]
Possibly caused [JENKINS-75255](https://issues.jenkins.io/browse/JENKINS-75255).

**Pull request merged**
